builtin/credential/aws/path_role.go

```diff
@@ -486,37 +541,41 @@ func (b *backend) pathRoleCreateUpdate(ctx context.Context, req *logical.Request
 	}
 
 	if boundIamRoleARNRaw, ok := data.GetOk("bound_iam_role_arn"); ok {
-		roleEntry.BoundIamRoleARN = boundIamRoleARNRaw.(string)
+		roleEntry.BoundIamRoleARNs = boundIamRoleARNRaw.([]string)
 	}
 
 	if boundIamInstanceProfileARNRaw, ok := data.GetOk("bound_iam_instance_profile_arn"); ok {
-		roleEntry.BoundIamInstanceProfileARN = boundIamInstanceProfileARNRaw.(string)
+		roleEntry.BoundIamInstanceProfileARNs = boundIamInstanceProfileARNRaw.([]string)
 	}
 
 	if boundIamPrincipalARNRaw, ok := data.GetOk("bound_iam_principal_arn"); ok {
-		principalARN := boundIamPrincipalARNRaw.(string)
-		roleEntry.BoundIamPrincipalARN = principalARN
-		// Explicitly not checking to see if the user has changed the ARN under us
-		// This allows the user to sumbit an update with the same ARN to force Vault
-		// to re-resolve the ARN to the unique ID, in case an entity was deleted and
-		// recreated
-		if roleEntry.ResolveAWSUniqueIDs && roleEntry.BoundIamPrincipalARN != "" && !strings.HasSuffix(roleEntry.BoundIamPrincipalARN, "*") {
-			principalID, err := b.resolveArnToUniqueIDFunc(ctx, req.Storage, principalARN)
-			if err != nil {
-				return logical.ErrorResponse(fmt.Sprintf("failed updating the unique ID of ARN %#v: %#v", principalARN, err)), nil
+		principalARNs := boundIamPrincipalARNRaw.([]string)
+		roleEntry.BoundIamPrincipalARNs = principalARNs
+		roleEntry.BoundIamPrincipalIDs = []string{}
+		for _, principalARN := range principalARNs {
+			// Explicitly not checking to see if the user has changed the ARN under us
+			// This allows the user to sumbit an update with the same ARN to force Vault
+			// to re-resolve the ARN to the unique ID, in case an entity was deleted and
+			// recreated
+			if roleEntry.ResolveAWSUniqueIDs && !strings.HasSuffix(principalARN, "*") {
+				principalID, err := b.resolveArnToUniqueIDFunc(ctx, req.Storage, principalARN)
+				if err != nil {
+					return logical.ErrorResponse(fmt.Sprintf("failed updating the unique ID of ARN %#v: %#v", principalARN, err)), nil
+				}
+				roleEntry.BoundIamPrincipalIDs = append(roleEntry.BoundIamPrincipalIDs, principalID)
 			}
-			roleEntry.BoundIamPrincipalID = principalID
-		} else {
-			// Need to handle the case where we're switching from a non-wildcard principal to a wildcard principal
-			roleEntry.BoundIamPrincipalID = ""
 		}
-	} else if roleEntry.ResolveAWSUniqueIDs && roleEntry.BoundIamPrincipalARN != "" && !strings.HasSuffix(roleEntry.BoundIamPrincipalARN, "*") {
-		// we're turning on resolution on this role, so ensure we update it
-		principalID, err := b.resolveArnToUniqueIDFunc(ctx, req.Storage, roleEntry.BoundIamPrincipalARN)
-		if err != nil {
-			return logical.ErrorResponse(fmt.Sprintf("unable to resolve ARN %#v to internal ID: %#v", roleEntry.BoundIamPrincipalARN, err)), nil
+	} else if roleEntry.ResolveAWSUniqueIDs && len(roleEntry.BoundIamPrincipalIDs) == 0 {
```

Comment:
I wonder if this should be an `else if` block or should it be an independent `if`. Don't we want to resolve the ARNs in the role anyways?